src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs

```diff
@@ -20,20 +28,106 @@ impl fmt::Debug for Address {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Default)]
 pub struct NetworkConfig {
+    pub index: u64,
     /// Host level path for the guest network interface.
     pub host_dev_name: String,
 
+    /// Guest iface name for the guest network interface.
+    pub virt_iface_name: String,
+
     /// Guest MAC address.
     pub guest_mac: Option<Address>,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Default)]
 pub struct NetworkDevice {
     /// Unique identifier of the device
-    pub id: String,
-
+    pub device_id: String,
+    pub attach_count: u64,
     /// Network Device config info
     pub config: NetworkConfig,
 }
+
+impl NetworkDevice {
+    // new creates a NetworkDevice
+    pub fn new(device_id: String, config: &NetworkConfig) -> Self {
+        Self {
+            device_id,
+            attach_count: 0,
+            config: config.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl Device for NetworkDevice {
+    async fn attach(&mut self, h: &dyn hypervisor) -> Result<()> {
+        if self
+            .increase_attach_count()
+            .await
+            .context("failed to increase attach count")?
+        {
+            return Ok(());
+        }
+
+        if let Err(e) = h.add_device(DeviceType::Network(self.clone())).await {
+            self.decrease_attach_count().await?;
+            return Err(e);
+        }
+
+        return Ok(());
+    }
+
+    async fn detach(&mut self, h: &dyn hypervisor) -> Result<Option<u64>> {
+        if self
+            .decrease_attach_count()
+            .await
+            .context("failed to decrease attach count")?
+        {
+            return Ok(None);
```

Comment:
It would be better if you leave some comments to explain why it returned early here.

Reply:
fixed